### PR TITLE
Harmony_timeseries - connects data from discrete time points

### PR DESCRIPTION
**Pull request by awnimo**
Harmony_timeseries, a framework for connecting scRNA-seq data from discrete time points

### CODE REVIEW

File: scanpy/external/tl/_harmony_timeseries.py [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+"""
+
+import pandas as pd
+from anndata import AnnData
+
+from ... import logging as logg
+
+
+def harmony_timeseries(data: AnnData, tp: str):
+    """Harmony time series for data visualization with augmented affinity matrix
```
Comment (flying-sheep) [inline scanpy/external/tl/_harmony_timeseries.py]:
```suggestion
    """\
    Harmony time series for data visualization with augmented affinity matrix
```

File: scanpy/external/tl/_harmony_timeseries.py [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+    at discrete time points [Nowotschin18i]_.
+
+    Harmony time series is a framework for data visualization, trajectory
+    detection and interpretation for scRNA-seq data measured at discrete
+    time points. Harmony constructs an augmented affinity matrix by augmenting
+    the kNN graph affinity matrix with mutually nearest neighbors between
+    successive time points. This augmented affinity matrix forms the basis for
+    generated a force directed layout for visualization and also serves as input
+    for computing the diffusion operator which can be used for trajectory
+    detection using **Palantir**.
```
Comment (flying-sheep) [inline scanpy/external/tl/_harmony_timeseries.py]: As said: Please don’t use anonymous links. Use links like this instead:

```suggestion
    detection using Palantir_.

    .. _Palantir: https://github.com/dpeerlab/Palantir
```

File: scanpy/external/tl/_harmony_timeseries.py [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            data_df=adata.to_df(),
+            timepoints=adata.obs["time_points"],
+            timepoint_connections=adata.uns["timepoint_connections"],
+        )
+
+    **Visualization using force directed layouts**
+
+    >>> layout = d.harmony_timeseries.plot.force_directed_layout(
+            affinity_matrix=aug_aff, cell_names=adata.obs.index
+        )
```
Comment (flying-sheep) [inline scanpy/external/tl/_harmony_timeseries.py]: Is there any reason you need to return a class instance? Why not

1. Do both of those steps within `sce.tl.harmony_timeseries`
2. store the resulting matrices within `adata`

It’s the way scanpy works, and unless there’s reasons not to, users will appreciate tools to follow that convention.

File: scanpy/external/tl/_harmony_timeseries.py [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+
+    Use any of Scanpy or Harmony methods for plotting. Example:
+
+    >>> d.harmony_timeseries.plot.plot_timepoints(
+            layout=layout, timepoints=adata.obs.time_points
+        )
```
Comment (flying-sheep) [inline scanpy/external/tl/_harmony_timeseries.py]: And if it’s possible to eliminate the class as I said above, the plotting function belongs into `sce.pl`

**Review by flying-sheep — CHANGES_REQUESTED** [review]
Thank you for addressing some of my concerns. There’s some issues remaining that you didn’t address, though, while marking the corresponding conversations as resolved.

Did you forget to push some commits or did you simply not get around to do them yet? In the second case, it would be nice to not mark conversations as resolved that you didn’t address yet.

File: scanpy/external/tl/_harmony_timeseries.py (hunk 4 above)
Comment (awnimo) [reply scanpy/external/tl/_harmony_timeseries.py]: Eliminated the class and reduced it into a simple function . Plotting can be done using `sc.pl.scatter` as illustrated in the documentation. Our harmony package has convenience plotting methods that perhaps we will add to `see.pl` in the future. In the meanwhile `sc.pl.scatter` does the same.

File: scanpy/external/tl/_harmony_timeseries.py (hunk 3 above)
Comment (awnimo) [reply scanpy/external/tl/_harmony_timeseries.py]: Done!

File: scanpy/external/tl/_harmony_timeseries.py (hunk 4 above)
Comment (flying-sheep) [reply scanpy/external/tl/_harmony_timeseries.py]: Great! Looks much simpler now. Thank you!

File: scanpy/external/tl/_harmony_timeseries.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        adata.to_df(),
+        adata.obs[tp],
+        adata.uns['timepoint_connections'],
+        pc_components=n_components,
+    )
+
+    # Force directed layouts
+    layout = harmony.plot.force_directed_layout(aug_aff, adata.obs.index)
+
+    adata.obs[['x', 'y']] = layout
```
Comment (flying-sheep) [inline scanpy/external/tl/_harmony_timeseries.py]: We’re storing embeddings like your layout in `.obsm["X_<embedding>"]`

```suggestion
    adata.obsm["X_harmony"] = layout
```

(Please adapt the example accordingly)

File: scanpy/external/tl/_harmony_timeseries.py [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+    adata.uns['aff'] = aff
+    adata.uns['aug_aff'] = aug_aff
```
Comment (flying-sheep) [inline scanpy/external/tl/_harmony_timeseries.py]: Which of those are specific to harmony, which ones are only dependent on the data? E.g. if `aff` is *the* affinity matrix for this object, and `aug_aff` is harmony specific, `aug_aff` should be stored as `aff_aug_harmony` or `aff_harmony`.

Heads-up: Starting with AnnData 0.7, which will be released soon, we’re storing n_obs×n_obs matrices in anndata’s `.obsp`. I’ll switch this over ASAP once AnnData 0.7 is out.

**Review by flying-sheep — CHANGES_REQUESTED** [review]
Great, this looks much simpler now, and very scanpy-like.

File: scanpy/external/tl/_harmony_timeseries.py (hunk 6 above)
Comment (awnimo) [reply scanpy/external/tl/_harmony_timeseries.py]: Actually they are both Harmony specific. I will change the names as suggested. When the release of AnnData 0.7 is out I will make the change to adapt the new data structure.

Comment (awnimo) [reply scanpy/external/tl/_harmony_timeseries.py]: @flying-sheep 
Please let me know if there is any other issues that need to be addressed.

**Review by flying-sheep — APPROVED** [review]

**Pull request merged**